### PR TITLE
Address Book API endpoints

### PR DESCRIPTION
**Pull request by kgorazd**

### CODE REVIEW

**Review by poudelprakash — APPROVED** [review]

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        module Account
+          class AddressesController < ::Spree::Api::V2::BaseController
+            before_action :require_spree_current_user
+
+            def index
+              render_serialized_payload { serialize_collection(collection) }
+            end
+
+            def create
+              resource = collection.new(address_params)
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:14]: @kgorazd let's move this into `Account::Addresses::Create` service

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+              if resource.save
+                render_serialized_payload { serialize_resource(resource) }
+              else
+                render_error_payload(Spree::ServiceModule::ResultError.new(resource.errors))
+              end
+            end
+
+            def update
+              if resource.update(address_params)
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:23]: and this to `Update` service

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            def address_params
+              replace_country_iso_with_id(params).require(:address).permit(permitted_address_attributes)
+            end
+
+            def replace_country_iso_with_id(params)
+              iso = params.dig(:address, :country_iso)
+              return params unless iso.present?
+
+              replaced_params = params
+              replaced_params[:address]['country_id'] = Spree::Country.find_by(iso: iso)&.id
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:61]:
```suggestion
              replaced_params[:address]['country_id'] = Spree::Country.by_iso(iso)&.id
```

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 1 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: 👍

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 2 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: 👍

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 3 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: 👍

Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 2 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 1 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        module Account
+          class AddressesController < ::Spree::Api::V2::BaseController
+            before_action :require_spree_current_user
+
+            def index
+              render_serialized_payload { serialize_collection(collection) }
+            end
+
+            def create
+              result = Spree::Account::Addresses::Create.call(user: spree_current_user, address_params: address_params)
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:14]: @kgorazd please use dependencies for this eg. https://github.com/spree/spree/blob/master/api/app/controllers/spree/api/v2/storefront/cart_controller.rb#L20

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+            def render_result(result)
+              if result.success?
+                render_serialized_payload { serialize_resource(result.value) }
+              else
+                render_error_payload(result.error)
+              end
+            end
+
+            def replace_country_iso_with_id(params)
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:57]: @kgorazd please move this part into Create / Update services - or extract this into a helper.

Also, you need to find `state_id` based on `state_name`

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 4 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 5 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+            def update
+              result = update_service.call(address: resource, address_params: address_params)
+              render_result(result)
+            end
+
+            private
+
+            def collection
+              spree_current_user.addresses
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:26]: @kgorazd let's move this into a finder class and also add it as a dependency, it's a common pain point (eg. filter only addresses per kind etc) eg. https://github.com/spree/spree/blob/966386c200175473d04d3449718fd985a6910474/api/app/controllers/spree/api/v2/storefront/countries_controller.rb#L32

File: core/app/services/spree/account/addresses/base.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        def fill_country_and_state_ids(params)
+          replace_country_iso_with_id(params)
+          fill_state_id(params)
+        end
+
+        def replace_country_iso_with_id(params)
+          iso = params[:country_iso]
+          return params unless iso.present?
+
+          params[:country_id] = Spree::Country.by_iso(iso)&.id
```
Comment (damianlegawiec) [inline core/app/services/spree/account/addresses/base.rb:18]: @kgorazd let's set `country` as an attribute of this class so we can refer to it later in `fill_state_id`

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+            def update
+              result = update_service.call(address: resource, address_params: address_params)
+              render_result(result)
+            end
+
+            private
+
+            def collection
+              scope = spree_current_user.addresses
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:26]: @kgorazd let's move `scope` into a seperate method, thanks!

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+              collection_finder.new(scope: scope, params: params).execute
+            end
+
+            def resource
+              @resource ||= collection.find(params[:id])
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb:31]:
```suggestion
              @resource ||= scope.find(params[:id])
```

Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb (hunk 8 above)
Comment (kgorazd) [reply api/app/controllers/spree/api/v2/storefront/account/addresses_controller.rb]: fixed

File: core/app/services/spree/account/addresses/create.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
+module Spree
+  module Account
+    module Addresses
+      class Create < Spree::Account::Addresses::Base
```
Comment (damianlegawiec) [inline core/app/services/spree/account/addresses/create.rb:4]:
```suggestion
      class Create < ::Spree::Account::Addresses::Base
```

File: core/app/services/spree/account/addresses/create.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        def call(user:, address_params:)
+          default_params = {
+            user: user
+          }
+          fill_country_and_state_ids(address_params)
+
+          address = Spree::Address.new(default_params.merge(address_params))
```
Comment (damianlegawiec) [inline core/app/services/spree/account/addresses/create.rb:11]:
```suggestion
          address = user.addresses.new(default_params.merge(address_params))
```

File: core/app/services/spree/account/addresses/create.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
+module Spree
+  module Account
+    module Addresses
+      class Create < Spree::Account::Addresses::Base
+        def call(user:, address_params:)
+          default_params = {
+            user: user
+          }
+          fill_country_and_state_ids(address_params)
```
Comment (damianlegawiec) [inline core/app/services/spree/account/addresses/create.rb:9]:
```suggestion
          address_params = fill_country_and_state_ids(address_params)
```

File: core/spec/services/spree/account/addresses/create_spec.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        address1: FFaker::Address.street_address,
+        city: FFaker::Address.city,
+        phone: FFaker::PhoneNumber.phone_number,
+        zipcode: FFaker::AddressUS.zip_code,
+        state_name: state.name,
+        country_iso: country.iso
+      }
+    end
+
+    describe '#call' do
```
Comment (damianlegawiec) [inline core/spec/services/spree/account/addresses/create_spec.rb:23]: @kgorazd let's also add scenarios for failure please

**Review by damianlegawiec — APPROVED** [review]

**Pull request merged**